penty/pentypes/builtins.py

```diff
@@ -472,21 +479,246 @@ def set_init(elts_ty=None):
+        iter_tys = Types[other_ty]['__iter__'](other_ty)
+        if not isinstance(iter_tys, set):
+            iter_tys = {iter_tys}
+        for iter_ty in iter_tys:
+            if '__next__' not in Types[iter_ty]:
+                raise TypeError
+            value_ty = Types[iter_ty]['__next__'](iter_ty)
+            intersection_tys.add(value_ty)
+    return _Set[intersection_tys]
+
+def set_intersection_update(self_ty, *other_tys):
+    if not issubclass(self_ty, set):
+        raise TypeError
+    updated = set_intersection(self_ty, *other_tys)
+    self_ty.__args__[0].update(updated.__args__[0])
+    return _Cst[None]
+
+def set_isdisjoint(self_ty, other_ty):
+    from penty.penty import Types
+    if not issubclass(self_ty, set):
+        raise TypeError
+    if not self_ty.__args__[0] or not other_ty.__args__[0]:
+        return _Cst[True]
+    if '__iter__' not in Types[other_ty]:
+        raise TypeError
+    return bool
+
+def set_issubset(self_ty, other_ty):
+    from penty.penty import Types
+    if not issubclass(self_ty, set):
+        raise TypeError
+    if not self_ty.__args__[0]:
+        return _Cst[True]
+    if '__iter__' not in Types[other_ty]:
+        raise TypeError
+    return bool
+
+def set_issuperset(self_ty, other_ty):
+    if not issubclass(self_ty, set):
+        raise TypeError
+    if not self_ty.__args__[0] and not other_ty.__args__[0]:
+        return _Cst[True]
+    return set_issubset(self_ty, other_ty)
+
+def set_pop(self_ty):
+    if not issubclass(self_ty, set):
+        raise TypeError
+    if not self_ty.__args__[0]:
+        raise TypeError
+    return self_ty.__args__[0].copy()
+
+def set_remove(self_ty, value_ty):
+    if not issubclass(self_ty, set):
+        raise TypeError
+    if not self_ty.__args__[0]:
+        raise TypeError
+    return _Cst[None]
+
+def set_symmetric_difference(self_ty, other_ty):
+    if not issubclass(self_ty, set):
+        raise TypeError
+    return set_intersection(self_ty, other_ty)
+
+def set_symmetric_difference_update(self_ty, other_ty):
+    if not issubclass(self_ty, set):
+        raise TypeError
+    return set_intersection_update(self_ty, other_ty)
+
+def set_union(self_ty, *other_tys):
+    if not issubclass(self_ty, set):
+        raise TypeError
+    return set_intersection(self_ty, *other_tys)
+
+def set_update(self_ty, *other_tys):
+    if not issubclass(self_ty, set):
+        raise TypeError
+    return set_intersection_update(self_ty, *other_tys)
+
 def set_instanciate(ty):
-    return {
-        '__bool__': _FT[lambda self_ty: bool],
-        '__and__': _FT[set_and],
-        '__contains__': _FT[lambda self_ty, value_ty: bool],
-        '__eq__': _FT[lambda self_ty, value_ty: bool],
-        '__ge__': make_set_compare(),
-        '__gt__': make_set_compare(),
-        '__iand__': _FT[set_iand],
-        '__len__': _FT[lambda self_ty: int],
-    }
+    return _set_methods
+
+_set_methods = {
+    '__bool__': _FT[lambda self_ty: bool],
```

Comment:
Here, you are probably missing a check that `issubclass(self_ty, ty)`

You could factorize this for multiple operations (`__contains__`, `__eq__`, `__ne__`):

```python
def set_bool_op(self_ty, *other_ty):
    if not isinstance(self_ty, set):
        raise TypeError
    return bool
```

For various reasons, there is not much we can say ever decide reguarding `*other_ty` (everything is always kind of valid).

The `__contains__` case is a bit disturbing since we would probably like to have

```python
def set_contains(self_ty, other_ty):
    […]
    if other_ty not in self_ty.__args__[0].
        raise TypeError
    return bool
```

but once again, the `1.0 in {1, 2}` is not helping…

Reply:
I had to specialize them because they have slightly different behavior wrt. corner cases.